### PR TITLE
sitting attributes for cat and parrot

### PR DESCRIPTION
**Pull request by UsainSrht**

### CODE REVIEW

**Comment by OliverSchlueter** [comment]
You can remove the cat sitting attribute, it is already added by another PR

**Comment by UsainSrht** [comment]
> You can remove the cat sitting attribute, it is already added by another PR

current cat sitting does trigger the sit event asynchronously

```[20:38:52 WARN]: [FancyNpcs] Plugin FancyNpcs v2.1.0-SNAPSHOT generated an exception while executing task 34234
java.lang.IllegalStateException: EntityToggleSitEvent may only be triggered synchronously.
        at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:42) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:627) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.event.Event.callEvent(Event.java:45) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at net.minecraft.world.entity.TamableAnimal.setInSittingPose(TamableAnimal.java:133) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
        at net.minecraft.world.entity.TamableAnimal.setInSittingPose(TamableAnimal.java:130) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
        at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.attributes.CatAttributes.setPose(CatAttributes.java:57) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
        at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.NpcAttribute.apply(NpcAttribute.java:37) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
        at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.NpcData.applyAllAttributes(NpcData.java:333) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
        at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.update(Npc_1_20_6.java:248) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
        at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.spawn(Npc_1_20_6.java:122) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
        at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.Npc.lambda$checkAndUpdateVisibility$1(Npc.java:114) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
        at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.20.6.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
[20:38:54 WARN]: java.lang.NoSuchFieldException: dataidFlags
[20:38:54 WARN]:        at java.base/java.lang.Class.getDeclaredField(Class.java:2782)
[20:38:54 WARN]:        at io.papermc.reflectionrewriter.runtime.AbstractDefaultRulesReflectionProxy.getDeclaredField(AbstractDefaultRulesReflectionProxy.java:90)
[20:38:54 WARN]:        at io.papermc.paper.pluginremap.reflect.PaperReflectionHolder.getDeclaredField(Unknown Source)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancylib.ReflectionUtils.getValue(ReflectionUtils.java:11)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.v1_20_6.attributes.PandaAttributes.setFlag(PandaAttributes.java:67)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.v1_20_6.attributes.PandaAttributes.setPose(PandaAttributes.java:52)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.api.NpcAttribute.apply(NpcAttribute.java:37)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.api.NpcData.applyAllAttributes(NpcData.java:333)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.update(Npc_1_20_6.java:248)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.spawn(Npc_1_20_6.java:122)
[20:38:54 WARN]:        at FancyNpcs-2.1.0-SNAPSHOT.jar//de.oliver.fancynpcs.api.Npc.lambda$checkAndUpdateVisibility$1(Npc.java:114)
[20:38:54 WARN]:        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:101)
[20:38:54 WARN]:        at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57)
[20:38:54 WARN]:        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22)
[20:38:54 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[20:38:54 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[20:38:54 WARN]:        at java.base/java.lang.Thread.run(Thread.java:1583)
[20:38:54 WARN]: [FancyNpcs] Plugin FancyNpcs v2.1.0-SNAPSHOT generated an exception while executing task 34297
java.lang.NullPointerException: null```

**Comment by OakLoaf** [comment]
> > You can remove the cat sitting attribute, it is already added by another PR
> 
> current cat sitting does trigger the sit event asynchronously
> 
> ```
> java.lang.IllegalStateException: EntityToggleSitEvent may only be triggered synchronously.
>         at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:42) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:627) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
>         at org.bukkit.event.Event.callEvent(Event.java:45) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
>         at net.minecraft.world.entity.TamableAnimal.setInSittingPose(TamableAnimal.java:133) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at net.minecraft.world.entity.TamableAnimal.setInSittingPose(TamableAnimal.java:130) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.attributes.CatAttributes.setPose(CatAttributes.java:57) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.NpcAttribute.apply(NpcAttribute.java:37) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.NpcData.applyAllAttributes(NpcData.java:333) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.update(Npc_1_20_6.java:248) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.spawn(Npc_1_20_6.java:122) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.Npc.lambda$checkAndUpdateVisibility$1(Npc.java:114) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.20.6.jar:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
>         at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
> ```

It might be worth putting this as an issue

**Comment by OliverSchlueter** [comment]
> > You can remove the cat sitting attribute, it is already added by another PR
> 
> current cat sitting does trigger the sit event asynchronously
> 
> ```
> java.lang.IllegalStateException: EntityToggleSitEvent may only be triggered synchronously.
>         at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:42) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:627) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
>         at org.bukkit.event.Event.callEvent(Event.java:45) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
>         at net.minecraft.world.entity.TamableAnimal.setInSittingPose(TamableAnimal.java:133) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at net.minecraft.world.entity.TamableAnimal.setInSittingPose(TamableAnimal.java:130) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.attributes.CatAttributes.setPose(CatAttributes.java:57) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.NpcAttribute.apply(NpcAttribute.java:37) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.NpcData.applyAllAttributes(NpcData.java:333) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.update(Npc_1_20_6.java:248) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.v1_20_6.Npc_1_20_6.spawn(Npc_1_20_6.java:122) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at FancyNpcs-2.1.0-SNAPSHOT.jar/de.oliver.fancynpcs.api.Npc.lambda$checkAndUpdateVisibility$1(Npc.java:114) ~[FancyNpcs-2.1.0-SNAPSHOT.jar:?]
>         at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.20.6.jar:1.20.6-86-9bf4855]
>         at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.20.6.jar:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
>         at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
> ```

I can't reproduce this? It works fine for me with 1.20.6

**Review by OliverSchlueter — APPROVED** [review]

**Comment by UsainSrht** [comment]
> I can't reproduce this? It works fine for me with 1.20.6

Works fine when set using commands but throws that when you go away and come back (load the npc)

**Comment by OliverSchlueter** [comment]
> > I can't reproduce this? It works fine for me with 1.20.6
> 
> Works fine when set using commands but throws that when you go away and come back (load the npc)

Ahh okay, I see. Need to set the data flag manually then

File: implementation_1_19_4/src/main/java/de/oliver/fancynpcs/v1_19_4/attributes/CatAttributes.java [hunk 1]
```diff
@@ -50,8 +50,21 @@ private static void setPose(Npc npc, String value) {
...
+            }
+            case "sleeping" -> {
+                cat.setOrderedToSit(false);
+                cat.setInSittingPose(false, false);
+                cat.setLying(true);
+            }
+            case "sitting" -> {
+                cat.setLying(false);
+                cat.setOrderedToSit(true);
+                cat.setInSittingPose(true, false);
```
Comment (UsainSrht) [inline implementation_1_19_4/src/main/java/de/oliver/fancynpcs/v1_19_4/attributes/CatAttributes.java]: second parameter is preventing the event from triggering so this pr also resolves that issue

**Comment by UsainSrht** [comment]
> Ahh okay, I see. Need to set the data flag manually then

`cat.setInSittingPose(true, false);` fixes it
Oak's pr was calling `setInSittingPose(boolean)` which also calls the event

**Comment by OliverSchlueter** [comment]
> > Ahh okay, I see. Need to set the data flag manually then
> 
> `cat.setInSittingPose(true, false);` fixes it Oak's pr was calling `setInSittingPose(boolean)` which also calls the event

Okay, I fixed it, thank you!

**Pull request merged**